### PR TITLE
postImagesCreate does actually produce JSON sequence (rfc7464).

### PR DESCRIPTION
**Pull request by ndeloof**
**- What I did**
Changed `Content-type` produced by `postImagesCreate` to adopt [RFC7464](https://datatracker.ietf.org/doc/rfc7464/) as an explicit mime type `application/json-seq`. The produced event stream is **not** a valid JSON object, so `application/json` should not be used.

**- How I did it**
Updated produced mime type

**- How to verify it**
curl -v --unix-socket /var/run/docker.sock -X POST http:/v1.41/images/create?fromImage=busybox

**- Description for the changelog**

Make it more explicit `images/create` produces JSON event sequence by adopting RFC7464 mime type `application/json-seq`.

**- A picture of a cute animal (not mandatory but encouraged)**
image

### CODE REVIEW

**Comment by thaJeztah** [comment]
Thanks! I seem to recall we made some changes in the content-type in the past that we had to revert as it broke some things. I'll have to dig into history for that though.

Perhaps it should perform content-negotiation, and depending on what the client advertises, use one or the other.

**Comment by ndeloof** [comment]
I also wondered about the potential impact changing the `Content-type`. Could probably only send this more relevant content type is client sends `Accept` header with explicit support for this mime type, wdyt?

**Comment by thaJeztah** [comment]
Yes, I was thinking along those lines indeed. When doing so, I think there's some utility package that we already use somewhere to make sure we parse `Accept` headers correctly (as they can be hard to parse)

These responses were always a bit "hairy" because when we implemented them, it was not really an official RFC yet, so there was no official content-type (other than some conventions used by some).

**Comment by ndeloof** [comment]
IIUC Kubernetes uses bitbucket.org/ww/goautoneg, so seems a reasonable dependency to be added to align with other projects in our ecosystem, wdyt?

**Comment by thaJeztah** [comment]
Thanks! So ... it looks like the bitbucket.org/ww/goautoneg package went AWOL, and they switch to a fork of it in https://github.com/kubernetes/kubernetes/commit/16fd72d6c91ba466a0e955a1d59a6c8d9e8791bc (switching to https://github.com/munnerz/goautoneg).

What _slightly_ concerns me is that it feels a bit like they "just picked a fork of the original one", which is in a personal repository (not an org) and the owner apparently just made a fork, but is not a real expert on the code itself https://github.com/munnerz/goautoneg/pull/3#pullrequestreview-298797987

> This seems good to me, although I'm not too well-versed in this code.

Not saying we _shouldn't_ switch to that package, but based on the above, I think it'd be good to have a good look how the implementation differs from what' we're currently using (if it's considerably better) before deciding whether or not we should switch. We could continue to use the current implementation (for now), and look for a possible replacement in a follow-up.

**Comment by ndeloof** [comment]
oh ok, I didn't checked the status of this dependency, makes perfect sense to keep our current implementation then, until the other ones demonstrates to be somehow a better one (?). Thanks for digging into this

**Comment by thaJeztah** [comment]
Thanks! We briefly went through this PR in the maintainers meeting, and we're good with this change in general.

I want to have another look, to look more into detail, but from an earlier glance (some of this could be in a follow-up);

* the swagger now only documents the new content-type as "produces" (I think it's possible to document both content-types separately, but I need to double check how swagger expects "content-negotiation" to be mentioned)
* there may be other endpoints that use the same "json-progress" output; we should probably make them all do the same (could be in a follow-up)
* we should check if the client code needs to be updated (i.e., make sure the client sends `Accepts` header for both `application/json` and the new content-type, so that it can talk with both "old" and "new" engines, and that new engines actually return the new content-type)

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!

**Comment by tianon** [comment]
Doesn't `application/json-seq` / RFC 7464 technically suggest the use of a "Record Separator" byte between JSON documents in addition to line-feed?  Aren't our APIs only using line-feed?

(Also, I'm not sure it matters technically, but the two swagger changes here appear to use different orderings in `produces:` that I think should probably be consistent.)

**Comment by ndeloof** [comment]
@tianon correct. I need to check how we can get `streamformatter.progressOutput` to not only append `LF` after progress messages but also prepend `RS`.

**Comment by thaJeztah** [comment]
Ah! Forgot to respond; good catch, @tianon (that _one time_ I didn't dive into the RFC 😬 😂).

Let me know if you need help, @ndeloof ! (although I'm not super familiar with the streamformatter either 😂)

**Comment by thaJeztah** [comment]
Before I forget; now that this PR needs updating, can you also add a note to the API history, and mention the new content-types / response types for these endpoints? https://github.com/moby/moby/blob/master/docs/api/version-history.md

File: pkg/streamformatter/streamformatter.go [hunk 1]
```diff
@@ -4,6 +4,8 @@ package streamformatter // import "github.com/docker/docker/pkg/streamformatter"
 import (
 	"encoding/json"
 	"fmt"
+	"github.com/docker/docker/api"
+	"github.com/docker/docker/api/types"
```
Comment (thaJeztah) [inline pkg/streamformatter/streamformatter.go:8]: Haven't looked very closely yet, but CI may fail on this (`pkg` package importing API code); perhaps we should do the reverse (define the const here, and the API consuming it)

(these imports should also be in the second group below)

Comment (ndeloof) [reply pkg/streamformatter/streamformatter.go:8]: so we have `api` to depend on `pkg`? I would expect the opposite is enforced

File: api/server/router/image/image_routes.go [hunk 2]
```diff
@@ -40,8 +41,10 @@ func (s *imageRouter) postImagesCreate(ctx context.Context, w http.ResponseWrite
 	)
 	defer output.Close()
 
-	w.Header().Set("Content-Type", httputil.NegotiateContentType(r,
-		[]string{"application/json-seq", "application/json"}, "application/json"))
+	mediaType := httputil.NegotiateContentType(r,
+		[]string{api.MediaTypeJsonSequence, "application/json"}, "application/json")
+	w.Header().Set("Content-Type", mediaType)
+	out := types.WithMediaType(output, mediaType)
```
Comment (thaJeztah) [inline api/server/router/image/image_routes.go:47]: Perhaps it makes sense to define a function or variable for this in the streamformatter (`streamformatter.MediaTypes`), as they'll now always be used together. (Perhaps we should even have a function that includes the negotiation, although not sure if that makes most sense in streamformater, or in the API 🤔)

File: api/swagger.yaml [hunk 3]
```diff
@@ -7634,7 +7634,7 @@ paths:
         - "text/plain"
         - "application/octet-stream"
       produces:
-        - "application/json"
+        - "application/json-seq"
```
Comment (thaJeztah) [inline api/swagger.yaml]: Does swagger allow specifying both content types here (like it does for `consumes`)?

**Review by thaJeztah — COMMENTED** [review]
LOL; had some comment pending it says

**Comment by ndeloof** [comment]
Closing as obsolete
